### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by VictorKunzler**
#20 
Extract buildFrame logic from canvas.
Remove Context.
Extract timer logic from canvas.
Remove buttonDefault, buttonFaster and buttonSlower components and create speedButton component.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,27 +1,50 @@
...
 import Nav from './components/nav';
 import Canvas from './components/canvas';
+import buildFrame from './utils/buildFrame';
+
+const FRAME_MAX_SIZE = 100;
 
 function App() {
-  const onState = useState(true);
-  const timerState = useState(1000);
+  const [ isOn, setIsOn] = useState(true);
```
Comment (nolleto) [inline src/App.js]:
```suggestion
  const [ isOn, setIsOn ] = useState(true);
```

File: src/utils/buildFrame.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	const B = Math.floor(Math.random() * 255).toString();
+	const color = `rgb(${R}, ${G}, ${B})`;
+
+	const size = Math.floor(Math.random() * maxSize);
+
+	const left = Math.floor(Math.random() * ( window.innerWidth - size ));
+	const bottom = Math.floor(Math.random() * ( window.innerHeight - size  ));
+
+	const newFrameStyle = {
+		color,
```
Comment (nolleto) [inline src/utils/buildFrame.js]: I believe this indentation is not right 🤔

File: src/__tests__/components/speedButton/speedButton.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            time={initialTime}
+            {...slowerButtonProps}
+          />
+        );
+
+        const { getByText } = render(button);
+
+        fireEvent.click(getByText('Slower'));
+
+        expect(initialTime).toBe(2000);
```
Comment (nolleto) [inline src/__tests__/components/speedButton/speedButton.js]: I think you could use `jest.fn()` here to detect the click event:
```js
it("set the time to double", () => {
  const onClick = jest.fn()
  const button = (
    <SpeedButton
      onClick={onClick}
      time={initialTime}
      {...slowerButtonProps}
    />
  );

  const { getByText } = render(button);

  fireEvent.click(getByText("Slower"));

  expect(onClick).toHaveBeenCalledWith(2000);
});
```

This way, you make sure the event `onClick` is been fired and also the value changes with `toHaveBeenCalledWith` method 😄 

Also, you can improve a little bit this test creating a method to render the component:
```js
const renderComponent = (props = {}) => render(<SpeedButton {...props} />);
```
Then, you can write you test like this:
```js
it("set the time to double", () => {
  const onClick = jest.fn();
  const { getByText } = renderComponent({
    onClick,
    time: initialTime,
    ...slowerButtonProps
  });

  fireEvent.click(getByText("Slower"));

  expect(onClick).toHaveBeenCalledWith(2000);
});
```

File: src/utils/buildFrame.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+const buildFrame = ({
+	maxSize,
+	frameFormat
+}) => {    
+	const R = Math.floor(Math.random() * 255).toString();
+	const G = Math.floor(Math.random() * 255).toString();
+	const B = Math.floor(Math.random() * 255).toString();
+	const color = `rgb(${R}, ${G}, ${B})`;
```
Comment (nolleto) [inline src/utils/buildFrame.js]: You can extract these to functions:

```js
const createRandomNumber = (max) => Math.floor(Math.random() * max)
const createRandomRGB = () =>
  `rgb(${createRandomNumber(255)}, ${createRandomNumber(255)}, ${createRandomNumber(255)})`;
```

And then, refactor `buildFrame` to: 
```js
const buildFrame = ({ maxSize, frameFormat }) => {
  const color = createRandomRGB();
  const size = createRandomNumber(maxSize);
  const left = createRandomNumber(window.innerWidth - size);
  const bottom = createRandomNumber(window.innerHeight - size);

  return {
    color,
    left,
    bottom,
    height: size,
    width: size,
    borderRadius: frameFormat === "circle" ? "50%" : "0%"
  };
};
```

File: src/utils/speedButtonsProps.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  },
+  {
+    text: 'Faster',
+    bgColor: 'green',
+    textColor: 'white',
+    speedMultiply: 0.5
+  },
+];
+
+export default speedButtonsProps;
```
Comment (nolleto) [inline src/utils/speedButtonsProps.js]: eof

Just let an empty line in the bottom of the file 😄

File: src/utils/buildFrame.js [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
-	const color = `rgb(${R}, ${G}, ${B})`;
-
-	const size = Math.floor(Math.random() * maxSize);
-
-	const left = Math.floor(Math.random() * ( window.innerWidth - size ));
-	const bottom = Math.floor(Math.random() * ( window.innerHeight - size  ));
+	const color = createRandomRGB();
+	const size = createRandomInt(maxSize);
+	const left = createRandomInt((window.innerWidth - size));
+	const bottom = createRandomInt((window.innerHeight - size));
```
Comment (nolleto) [inline src/utils/buildFrame.js]: You remove these double parentheses:
```diff
- const left = createRandomInt((window.innerWidth - size));
+ const left = createRandomInt(window.innerWidth - size);
```

💅

File: src/utils/createRandomRGB.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const createRandomRGB = () => {
+	const R = Math.floor(Math.random() * 255).toString();
+	const G = Math.floor(Math.random() * 255).toString();
+	const B = Math.floor(Math.random() * 255).toString();
```
Comment (nolleto) [inline src/utils/createRandomRGB.js]: You don't need to use `.toString()` here and you can use `createRandomInt`:
```js
const R = createRandomInt(255);
const G = createRandomInt(255);
const B = createRandomInt(255);
```
or
```js
const createRandomRGB = () =>
  `rgb(${createRandomInt(255)}, ${createRandomInt(255)}, ${createRandomInt(255)})`;
```

File: src/utils/speedButtonsProps.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const speedButtonsProps = [
+  {
+    text: 'Slower',
+    bgColor: 'red',
+    textColor: 'white',
+    speedMultiply: 2,
```
Comment (nolleto) [inline src/utils/speedButtonsProps.js]: Maybe `speedMultiplier` 🤔

**Comment by VictorKunzler** [comment]
@MEHDI-ROBATJAZI hey, can you merge this PR?

**Pull request closed without merging**